### PR TITLE
Adds 'close to tray' option in settings

### PR DESCRIPTION
**Pull request by plafue**
_**WIP**_
(see https://www.keepassx.org/dev/issues/364#change-1035)

I'd like to add this functionality, however i can not manually test this, as it seems like assets like icons and images are not properly loaded when i run the compiled version (e.g. i see the alternative caption "Add new entry" in the menu bar instead of its icon). The try icon is not displayed at all.

I'm not that familiar with C/Qt development, so any hints are very much appreciated.

Good work, by the way!

### CODE REVIEW

**Comment by nmschulte** [comment]
I just tested this myself and it seems to work well.  @debfx, what can I do to help get this in the next release?

File: src/gui/MainWindow.cpp [hunk 1]
```diff
@@ -431,6 +431,13 @@ void MainWindow::databaseTabChanged(int tabIndex)
 
 void MainWindow::closeEvent(QCloseEvent* event)
 {
+
+  if (config()->get("GUI/CloseToTray").toBool()) {
+    event->ignore();
+    QTimer::singleShot(0, this, SLOT(hide()));
```
Comment (nmschulte) [inline src/gui/MainWindow.cpp]: I've never worked with QT before.  Is the `singleShot` `hide` to release the event thread?

**Comment by kajzersoze** [comment]
Working fine, but how do you actually close the application if you want it ?
Right click on tray -> Quit doesn't work, same happens when selecting Quit from the Menu.

**Comment by nmschulte** [comment]
Send the process a SIGTERM or SIGKILL signal; that's the only way to close it with this change and the feature enabled.

**Comment by kajzersoze** [comment]
Well yeah, anyway I added something to quit with a middle click on tray icon and it works.

src/gui/MainWindow.cpp
```
void MainWindow::trayIconTriggered(QSystemTrayIcon::ActivationReason reason)
 {
     if (reason == QSystemTrayIcon::Trigger) {
         toggleWindow();
    }
+    if (reason == QSystemTrayIcon::MiddleClick) {
+         reallyQuit();
+     }
 }


 
+void MainWindow::reallyQuit()
+{
+    bool accept = saveLastDatabases();
+
+    if (accept) {
+        saveWindowInformation();
+        QApplication::quit();
+    }
+    else {
+      QApplication::quit();
+    }
+}
+
 void MainWindow::lockDatabasesAfterInactivity()
 {

```

src/gui/MainWindow.h

```
     void toggleWindow();
+    void reallyQuit();
     void lockDatabasesAfterInactivity();
```

**Comment by nmschulte** [comment]
I think choosing the "Quit" option from the tray context-menu should actually kill the program; when would you ever want to use that (remember, from the tray...) to minimize the program?  If someone did want the minimize action, it'd probably be better to add a "Minimize" entry.

File: src/gui/SettingsWidgetGeneral.ui [hunk 2]
```diff
@@ -113,6 +113,16 @@
      </property>
     </widget>
    </item>
+   <item row="12" column="0">
+    <widget class="QCheckBox" name="systrayCloseToTrayCheckBox">
+     <property name="enabled">
+      <bool>false</bool>
+     </property>
+     <property name="text">
+      <string>Hide window to system tray when closing main window (Use Ctrl+Q or Database -> Exit to exit).</string>
```
Comment (nmschulte) [inline src/gui/SettingsWidgetGeneral.ui]: @kajzersoze, I see what you're getting at now.  This text will definitely need to change to match whatever the decided behavior is: it doesn't match even the patch it comes with.

**Comment by kajzersoze** [comment]
Yeah exactly, that's how it should be.
I just didn't know how to do that (quit from the tray -> reallyQuit) so I added middle click thingy, which works from me atm until someone smarter figure out the real Quit from the tray :)

**Comment by nmschulte** [comment]
I'll try and clean this up and see if I can get it in.  @plafue, are you still wanting to see this included too?

**Comment by kajzersoze** [comment]
I figured out actually how to add an option to the tray and it works.
Now probably good idea would be to remove Quit from the tray menu, not sure though what to do with the Quit option from Main Menu. Middle click on tray is also not needed now.
Here's the addition to the tray menu, needs reallyQuit() from earlier.
src/gui/MainWindow.cpp
```
             QAction* actionToggle = new QAction(tr("Toggle window"), menu);
             menu->addAction(actionToggle);
+	    QAction* actionToggle = new QAction(tr("Really Quit"), menu);
+            menu->addAction(actionReallyQuit);	
 
             menu->addAction(m_ui->actionQuit);
 
             connect(m_trayIcon, SIGNAL(activated(QSystemTrayIcon::ActivationReason)),
                     SLOT(trayIconTriggered(QSystemTrayIcon::ActivationReason)));
             connect(actionToggle, SIGNAL(triggered()), SLOT(toggleWindow()));
+	    connect(actionReallyQuit, SIGNAL(triggered()), SLOT(reallyQuit()));
 
             m_trayIcon->setContextMenu(menu);
             m_trayIcon->show();
```

**Pull request closed without merging**